internal/common/normalattackstate.go

```diff
@@ -134,3 +140,146 @@ func Get5PercentN0Delay(activeChar *character.CharWrapper) int {
+			}
+		}
+
+		return false
+	}, fmt.Sprintf("%v animation check", w.abilName))
+}
+
+func (w *NAHook) naStateDelayFuncGen(src int) func() {
+	return func() {
+		//ignore if on ICD
+		if w.c.StatusIsActive(w.abilICDKey) || w.Core.Player.CurrentState() != action.NormalAttackState || w.abilHookSrc != src {
+			w.Core.Log.NewEvent(fmt.Sprintf("%v did not trigger on state change", w.abilName), glog.LogCharacterEvent, w.c.Index).
+				Write("state", w.Core.Player.CurrentState()).
+				Write("icd", w.c.StatusExpiry(w.abilICDKey)).
+				Write("ICD check", w.c.StatusIsActive(w.abilICDKey)).
+				Write("State check", w.Core.Player.CurrentState() != action.NormalAttackState).
+				Write("Source check", w.abilHookSrc != src)
+
+			return
+		}
+		//this should start a new ticker if not on ICD and state is correct
+		w.summonFunc()
+		w.c.AddStatus(w.abilICDKey, 60, true)
+		w.Core.Log.NewEvent(fmt.Sprintf("%v on state change", w.abilName), glog.LogCharacterEvent, w.c.Index).
+			Write("state", action.NormalAttackState).
+			Write("icd", w.c.StatusExpiry(w.abilICDKey))
+		w.abilTickSrc = w.Core.F
+		//use the hitlag affected queue for this
+		w.c.QueueCharTask(w.naTickerFunc(w.Core.F), w.abilProcICD) //check every `ICD` frames
+	}
+}
+
+func (w *NAHook) naTickerFunc(src int) func() {
+	return func() {
+		//check if buff is up
+		if !w.c.StatusIsActive(w.abilKey) {
+			return
+		}
+		if w.abilTickSrc != src {
+			w.Core.Log.NewEvent(fmt.Sprintf("%v tick check ignored, src diff", w.abilName), glog.LogCharacterEvent, w.c.Index).
+				Write("src", src).
+				Write("new src", w.abilTickSrc)
+			return
+		}
+		//stop if we are no longer in normal animation state
+		state := w.Core.Player.CurrentState()
+
+		if state != action.NormalAttackState {
+			w.Core.Log.NewEvent(fmt.Sprintf("%v tick check stopped, not in normal state", w.abilName), glog.LogCharacterEvent, w.c.Index).
+				Write("src", src).
+				Write("state", state)
+			return
+		}
+		state_start := w.Core.Player.CurrentStateStart()
+		norm_counter := w.Core.Player.ActiveChar().NormalCounter
+		if (norm_counter == 1) && w.Core.F-state_start < w.delayFunc(w.Core.Player.ActiveChar()) {
+			w.Core.Log.NewEvent(fmt.Sprintf("%v tick check stopped, not enough time since normal state start", w.abilName), glog.LogCharacterEvent, w.c.Index).
+				Write("src", src).
+				Write("state_start", state_start)
+			return
+		}
+
+		w.Core.Log.NewEvent(fmt.Sprintf("%v triggered from ticker", w.abilName), glog.LogCharacterEvent, w.c.Index).
+			Write("src", src).
+			Write("state", state).
+			Write("icd", w.c.StatusExpiry(w.abilICDKey))
+
+		//we can trigger here b/c we're in normal state still and src is still the same
+		w.summonFunc()
+		w.c.AddStatus(w.abilICDKey, 60, true)
+		//in theory this should not hit an icd?
+		//use the hitlag affected queue for this
+		w.abilTickSrc = w.Core.F
+		w.c.QueueCharTask(w.naTickerFunc(w.Core.F), w.abilProcICD) //check every 1sec
```

Comment:
Same comment with the other two places. Seems like this log + 4 operations can be reused